File: bindings/julia/src/exceptions.jl
struct DatabaseException <: Exception
    msg::String
end

quiver_database_sqlite_error(msg::String) = throw(DatabaseException(msg))
Base.showerror(io::IO, e::DatabaseException) = print(io, e.msg)

"""
    check_error(err::C.quiver_error_t)

Check for C API errors and throw DatabaseException with the message from `quiver_get_last_error()`.
"""
function check_error(err)
    if err != C.QUIVER_OK
        detail = unsafe_string(C.quiver_get_last_error())
        if isempty(detail)
            @warn "check_error: C API returned error but quiver_get_last_error() is empty"
            throw(DatabaseException("Unknown error"))
        end
        throw(DatabaseException(detail))
    end
    return nothing
end

File: bindings/dart/lib/src/exceptions.dart
import 'package:ffi/ffi.dart';

import 'ffi/library_loader.dart';

/// Check for C API errors and throw the appropriate DatabaseException.
void check(int err) {
  if (err != 0) {
    final detail = bindings.quiver_get_last_error().cast<Utf8>().toDartString();
    if (detail.isEmpty) {
      print('WARNING check: C API returned error but quiver_get_last_error() is empty');
      throw const DatabaseException('Unknown error');
    }
    throw DatabaseException(detail);
  }
}

/// Base exception for Quiver errors.
class DatabaseException implements Exception {
  final String message;
  const DatabaseException(this.message);

  @override
  String toString() => 'DatabaseException: $message';
}

class LuaException extends DatabaseException {
  const LuaException(super.message);
}
